File: src/timeline.ts
import { randomUUID } from "node:crypto";

import { captureCurrentState, type CurrentState } from "./current-state.js";
import { diffGraphs, type GraphDiff, type SemanticChange } from "./diff.js";
import {
  appendTimelineEntry,
  latestTimelineEntry,
  readSnapshot,
  writeSnapshot
} from "./store.js";
import type { RuntimeOptions } from "./cli-shared.js";
import type { AgentId, TimelineChangedSurface, TimelineEntry, TimelineRestoreReadiness } from "./types.js";

export interface CaptureTimelineOptions {
  daemonRunId: string;
  snapshotName?: string;
  title?: string;
  skipUnchanged?: boolean;
}

export interface CaptureTimelineResult {
  written: boolean;
  entry?: TimelineEntry;
  state: CurrentState;
  diff?: GraphDiff;
  skippedReason?: "unchanged";
}

export async function captureTimelineSnapshot(
  options: RuntimeOptions,
  captureOptions: CaptureTimelineOptions
): Promise<CaptureTimelineResult> {
  const previous = await latestTimelineEntry(options.storeDir, {
    projectPath: options.projectPath,
    agent: options.agent
  });
  const snapshotName = captureOptions.snapshotName ?? timelineSnapshotName(captureOptions.daemonRunId, options.agent);
  const state = await captureCurrentState(options, snapshotName);

  let diff: GraphDiff | undefined;
  let diffError: string | undefined;
  if (previous) {
    try {
      const previousSnapshot = await readSnapshot(options.storeDir, previous.afterSnapshotName, previous.agent);
      diff = diffGraphs(previousSnapshot.graph, state.snapshot.graph);
    } catch (error) {
      diffError = error instanceof Error ? error.message : String(error);
      diff = undefined;
    }
  }

  if (
    previous &&
    diff &&
    captureOptions.skipUnchanged &&
    diff.semanticChanges.length === 0 &&
    diff.rawSourceChanges.length === 0
  ) {
    return { written: false, state, diff, skippedReason: "unchanged" };
  }

  const changedSurfaces = changedSurfacesForDiff(diff);
  const restoreReadiness = restoreReadinessFor(changedSurfaces);
  const entry: TimelineEntry = {
    schemaVersion: "0.1",
    id: shortId(),
    source: "daemon",
    eventKind: eventKindFor(previous, diff),
    title: captureOptions.title ?? titleForDiff(diff, options.agent),
    projectPath: state.snapshot.manifest.projectPath,
    ...(options.agent ? { agent: options.agent } : {}),
    agents: agentsForState(state),
    ...(previous ? { beforeSnapshotName: previous.afterSnapshotName } : {}),
    afterSnapshotName: snapshotName,
    daemonRunId: captureOptions.daemonRunId,
    createdAt: new Date().toISOString(),
    observedAt: state.snapshot.manifest.createdAt,
    changedSurfaces,
    restoreReadiness,
    confidence: confidenceFor(diff, changedSurfaces, diffError),
    confidenceReason: confidenceReasonFor(diff, changedSurfaces, diffError),
    evidenceCount: state.snapshot.evidence.length,
    graphNodeCount: state.snapshot.graph.length,
    auditFindingCount: state.snapshot.auditFindings.length,
    changes: {
      ...(previous ? { previousEntryId: previous.id, previousSnapshotName: previous.afterSnapshotName } : {}),
      hasChanges: !diff || diff.semanticChanges.length > 0 || diff.rawSourceChanges.length > 0,
      semanticChangeCount: diff?.semanticChanges.length ?? 0,
      rawSourceChangeCount: diff?.rawSourceChanges.length ?? 0,
      highlights: highlightsForDiff(diff)
    }
  };

  await writeSnapshot(options.storeDir, state.snapshot, options.agent);
  await appendTimelineEntry(options.storeDir, entry);

  return { written: true, entry, state, diff };
}

export function timelineSnapshotName(runId: string, agent?: AgentId): string {
  const timestamp = new Date().toISOString().replace(/[:.]/g, "-");
  return ["daemon", runId, agent, timestamp, shortId()].filter(Boolean).join("-");
}

function shortId(): string {
  return randomUUID().replace(/-/g, "").slice(0, 8);
}

function agentsForState(state: CurrentState): AgentId[] {
  return [...new Set(state.snapshot.evidence.map((item) => item.agent))].sort();
}

function titleForDiff(diff: GraphDiff | undefined, agent?: AgentId): string {
  if (!diff) {
    return scopedTitle("baseline setup", agent);
  }

  if (diff.semanticChanges.length === 0 && diff.rawSourceChanges.length === 0) {
    return scopedTitle("unchanged setup", agent);
  }

  const first = [...diff.semanticChanges].sort((left, right) =>
    priorityForChange(left).localeCompare(priorityForChange(right))
  )[0];

  if (!first) {
    return scopedTitle("update setup files", agent);
  }

  switch (first.code) {
    case "MCP_ADDED":
      return scopedTitle(`add ${first.entityName} mcp`, agent);
    case "MCP_REMOVED":
      return scopedTitle(`remove ${first.entityName} mcp`, agent);
    case "MCP_CHANGED":
      return scopedTitle(`update ${first.entityName} mcp`, agent);
    case "ENV_KEY_ADDED":
      return scopedTitle(`add ${first.entityName} env key`, agent);
    case "ENV_KEY_REMOVED":
      return scopedTitle(`remove ${first.entityName} env key`, agent);
    case "PERMISSION_WILDCARD_ADDED":
      return scopedTitle("update permissions", agent);
    case "SKILL_EXECUTABLE_APPEARED":
      return scopedTitle(`update ${first.entityName} skill`, agent);
    case "UNSUPPORTED_STATE_CHANGED":
      return scopedTitle("update unsupported setup", agent);
  }
}

function changedSurfacesForDiff(diff: GraphDiff | undefined): TimelineChangedSurface[] {
  if (!diff) return [];

  const surfaces: TimelineChangedSurface[] = [];
  const semanticSourcePaths = new Set<string>();
  for (const change of diff.semanticChanges) {
    const kind = timelineSurfaceKind(change);
    const path = typeof change.details.sourcePath === "string" ? change.details.sourcePath : "unknown";
    semanticSourcePaths.add(path);
    const restorable = kind === "mcp_server" && path.endsWith(".mcp.json");
    surfaces.push({
      kind,
      changeType: change.code,
      path,
      entityName: change.entityName,
      restorable,
      observeOnly: !restorable,
      ...(change.before === undefined ? {} : { before: change.before }),
      ...(change.after === undefined ? {} : { after: change.after })
    });
  }

  for (const change of diff.rawSourceChanges) {
    if (semanticSourcePaths.has(change.sourcePath)) continue;
    surfaces.push({
      kind: "other",
      changeType: `RAW_${change.status.toUpperCase()}`,
      path: change.sourcePath,
      restorable: false,
      observeOnly: true
    });
  }
  return surfaces;
}

function eventKindFor(previous: TimelineEntry | undefined, diff: GraphDiff | undefined): TimelineEntry["eventKind"] {
  if (!previous) return "baseline";
  if (diff && diff.semanticChanges.length === 0 && diff.rawSourceChanges.length === 0) return "unchanged";
  return "setup_changed";
}

function timelineSurfaceKind(change: SemanticChange): TimelineChangedSurface["kind"] {
  if (change.entityKind === "mcp_server") return "mcp_server";
  if (change.entityKind === "skill") return "skill";
  if (change.entityKind === "permission") return "permission";
  if (change.entityKind === "hook") return "hook";
  if (change.entityKind === "env_key") return "env_key";
  if (change.entityKind === "unsupported") return "unsupported";
  return "other";
}

function restoreReadinessFor(surfaces: TimelineChangedSurface[]): TimelineRestoreReadiness {
  if (surfaces.length === 0) return "observe-only";
  const restorable = surfaces.filter((surface) => surface.restorable).length;
  if (restorable === surfaces.length) return "full";
  if (restorable > 0) return "partial";
  return "observe-only";
}

function confidenceFor(diff: GraphDiff | undefined, surfaces: TimelineChangedSurface[], diffError?: string): TimelineEntry["confidence"] {
  if (diffError) return "low";
  if (!diff) return "high";
  if (surfaces.some((surface) => surface.path === "unknown")) return "medium";
  return "high";
}

function confidenceReasonFor(diff: GraphDiff | undefined, surfaces: TimelineChangedSurface[], diffError?: string): string {
  if (diffError) return `previous snapshot could not be diffed: ${diffError}`;
  if (!diff) return "first daemon baseline";
  if (surfaces.length === 0) return "no semantic or raw source changes";
  if (surfaces.some((surface) => surface.path === "unknown")) return "some changes lacked source path metadata";
  return "derived from snapshot graph diff";
}

function priorityForChange(change: SemanticChange): string {
  if (change.code.startsWith("MCP_")) return `0-${change.entityName}`;
  if (change.code === "SKILL_EXECUTABLE_APPEARED") return `1-${change.entityName}`;
  if (change.code === "PERMISSION_WILDCARD_ADDED") return `2-${change.entityName}`;
  if (change.code.startsWith("ENV_KEY_")) return `4-${change.entityName}`;
  return `5-${change.entityName}`;
}

function scopedTitle(title: string, agent?: AgentId): string {
  return agent ? `${title} for ${agentLabel(agent)}` : title;
}

function agentLabel(agent: AgentId): string {
  if (agent === "claude-code") return "Claude Code";
  if (agent === "codex") return "Codex";
  if (agent === "cursor") return "Cursor";
  if (agent === "opencode") return "OpenCode";
  if (agent === "pi-agent") return "Pi Agent";
  if (agent === "project") return "Project";
  return "Unknown";
}

function highlightsForDiff(diff: GraphDiff | undefined): string[] {
  if (!diff) return [];

  const highlights: string[] = [];
  for (const change of diff.semanticChanges.slice(0, 5)) {
    highlights.push(`${change.code}: ${change.entityName}`);
  }
  for (const change of diff.rawSourceChanges.slice(0, Math.max(0, 8 - highlights.length))) {
    highlights.push(`${change.status}: ${change.sourcePath}`);
  }
  return highlights;
}

File: crates/hem-core/src/timeline.rs
use std::collections::HashSet;
use std::path::Path;

use time::format_description::well_known::Rfc3339;
use time::OffsetDateTime;
use uuid::Uuid;

use crate::diff::{diff_graphs, GraphDiff, SemanticChange, SemanticChangeCode};
use crate::snapshot::capture_current_state;
use crate::store::{
    append_timeline_entry, latest_timeline_entry, read_snapshot, write_snapshot, StoreError,
    StoreSnapshot,
};
use crate::types::{
    AgentId, CurrentState, EvidenceKind, RuntimeOptions, TimelineChangeSummary,
    TimelineChangedSurface, TimelineConfidence, TimelineEntry, TimelineEntryEventKind,
    TimelineEntrySource, TimelineRestoreReadiness,
};

#[derive(Debug, thiserror::Error)]
pub enum TimelineError {
    #[error("Timeline entry not found: {0}")]
    NotFound(String),
    #[error(transparent)]
    Store(#[from] StoreError),
    #[error(transparent)]
    Io(#[from] std::io::Error),
}

#[derive(Debug, Clone, Default)]
pub struct CaptureTimelineOptions {
    pub capture_id: Option<String>,
    pub snapshot_name: Option<String>,
    pub title: Option<String>,
    pub skip_unchanged: bool,
}

#[derive(Debug, Clone)]
pub struct CaptureTimelineResult {
    pub written: bool,
    pub entry: Option<TimelineEntry>,
    pub state: CurrentState,
    pub diff: Option<GraphDiff>,
    pub skipped_reason: Option<String>,
}

pub fn capture_timeline_snapshot(
    options: &RuntimeOptions,
    capture_options: &CaptureTimelineOptions,
) -> Result<CaptureTimelineResult, TimelineError> {
    let store_dir = Path::new(&options.store_dir);
    let previous = latest_timeline_entry(
        store_dir,
        crate::store::TimelineListOptions {
            agent: options.agent,
            project_path: Some(&options.project_path),
            limit: None,
            on_corrupt_entry: None,
        },
    )?;

    let capture_id = capture_options
        .capture_id
        .clone()
        .unwrap_or_else(short_id);
    let snapshot_name = capture_options
        .snapshot_name
        .clone()
        .unwrap_or_else(|| timeline_snapshot_name(&capture_id, options.agent));
    let state = capture_current_state(options, &snapshot_name)?;

    let mut diff: Option<GraphDiff> = None;
    let mut diff_error: Option<String> = None;
    if let Some(previous) = &previous {
        match read_snapshot(store_dir, &previous.after_snapshot_name, previous.agent) {
            Ok(previous_snapshot) => {
                diff = Some(diff_graphs(
                    &previous_snapshot.graph,
                    &state.snapshot.graph,
                ));
            }
            Err(error) => {
                diff_error = Some(error.to_string());
            }
        }
    }

    if let (Some(_), Some(ref diff)) = (&previous, &diff) {
        if capture_options.skip_unchanged
            && diff.semantic_changes.is_empty()
            && diff.raw_source_changes.is_empty()
        {
            return Ok(CaptureTimelineResult {
                written: false,
                entry: None,
                state,
                diff: Some(diff.clone()),
                skipped_reason: Some("unchanged".to_string()),
            });
        }
    }

    let changed_surfaces = changed_surfaces_for_diff(diff.as_ref());
    let restore_readiness = restore_readiness_for(&changed_surfaces);
    let entry = TimelineEntry {
        schema_version: "0.1".to_string(),
        id: short_id(),
        source: TimelineEntrySource::Manual,
        event_kind: event_kind_for(previous.as_ref(), diff.as_ref()),
        title: capture_options
            .title
            .clone()
            .unwrap_or_else(|| title_for_diff(diff.as_ref(), options.agent)),
        project_path: state.snapshot.manifest.project_path.clone(),
        agent: options.agent,
        agents: agents_for_state(&state),
        before_snapshot_name: previous.as_ref().map(|p| p.after_snapshot_name.clone()),
        after_snapshot_name: snapshot_name.clone(),
        capture_id: capture_id.clone(),
        created_at: OffsetDateTime::now_utc()
            .format(&Rfc3339)
            .unwrap_or_else(|_| "1970-01-01T00:00:00Z".to_string()),
        observed_at: state.snapshot.manifest.created_at.clone(),
        changed_surfaces: changed_surfaces.clone(),
        restore_readiness,
        confidence: confidence_for(diff.as_ref(), &changed_surfaces, diff_error.as_deref()),
        confidence_reason: confidence_reason_for(diff.as_ref(), &changed_surfaces, diff_error.as_deref()),
        evidence_count: state.snapshot.evidence.len() as u32,
        graph_node_count: state.snapshot.graph.len() as u32,
        audit_finding_count: state.snapshot.audit_findings.len() as u32,
        changes: TimelineChangeSummary {
            previous_entry_id: previous.as_ref().map(|p| p.id.clone()),
            previous_snapshot_name: previous.as_ref().map(|p| p.after_snapshot_name.clone()),
            has_changes: diff.as_ref().map_or(true, |d| {
                !d.semantic_changes.is_empty() || !d.raw_source_changes.is_empty()
            }),
            semantic_change_count: diff.as_ref().map_or(0, |d| d.semantic_changes.len() as u32),
            raw_source_change_count: diff.as_ref().map_or(0, |d| d.raw_source_changes.len() as u32),
            highlights: highlights_for_diff(diff.as_ref()),
        },
    };

    write_snapshot(store_dir, StoreSnapshot::from(state.snapshot.clone()), options.agent)?;
    append_timeline_entry(store_dir, &entry)?;

    Ok(CaptureTimelineResult {
        written: true,
        entry: Some(entry),
        state,
        diff,
        skipped_reason: None,
    })
}

pub fn timeline_snapshot_name(capture_id: &str, agent: Option<AgentId>) -> String {
    let timestamp = OffsetDateTime::now_utc()
        .format(&Rfc3339)
        .unwrap_or_else(|_| "1970-01-01T00:00:00Z".to_string())
        .replace([':', '.'], "-");
    [
        Some("history".to_string()),
        Some(capture_id.to_string()),
        agent.map(|a| a.as_str().to_string()),
        Some(timestamp),
        Some(short_id()),
    ]
    .into_iter()
    .flatten()
    .collect::<Vec<_>>()
    .join("-")
}

fn short_id() -> String {
    Uuid::new_v4().to_string().replace('-', "")[..8].to_string()
}

fn agents_for_state(state: &CurrentState) -> Vec<AgentId> {
    let mut agents: Vec<AgentId> = state
        .snapshot
        .evidence
        .iter()
        .map(|item| item.agent)
        .collect::<HashSet<_>>()
        .into_iter()
        .collect();
    agents.sort_by_key(|a| a.as_str().to_string());
    agents
}

fn title_for_diff(diff: Option<&GraphDiff>, agent: Option<AgentId>) -> String {
    let Some(diff) = diff else {
        return scoped_title("baseline setup", agent);
    };

    if diff.semantic_changes.is_empty() && diff.raw_source_changes.is_empty() {
        return scoped_title("unchanged setup", agent);
    }

    let mut sorted = diff.semantic_changes.clone();
    sorted.sort_by(|left, right| {
        priority_for_change(left)
            .cmp(&priority_for_change(right))
    });

    let Some(first) = sorted.first() else {
        return scoped_title("update setup files", agent);
    };

    let title = match first.code {
        SemanticChangeCode::AgentConfigAdded
        | SemanticChangeCode::AgentConfigRemoved
        | SemanticChangeCode::AgentConfigChanged => "update config",
        SemanticChangeCode::McpAdded => return scoped_title(&format!("add {} mcp", first.entity_name), agent),
        SemanticChangeCode::McpRemoved => {
            return scoped_title(&format!("remove {} mcp", first.entity_name), agent);
        }
        SemanticChangeCode::McpChanged => {
            return scoped_title(&format!("update {} mcp", first.entity_name), agent);
        }
        SemanticChangeCode::EnvKeyAdded => {
            return scoped_title(&format!("add {} env key", first.entity_name), agent);
        }
        SemanticChangeCode::EnvKeyRemoved => {
            return scoped_title(&format!("remove {} env key", first.entity_name), agent);
        }
        SemanticChangeCode::PermissionWildcardAdded | SemanticChangeCode::PermissionChanged => {
            "update permissions"
        }
        SemanticChangeCode::SkillAdded => {
            return scoped_title(&format!("install {} skill", first.entity_name), agent);
        }
        SemanticChangeCode::SkillRemoved => {
            return scoped_title(&format!("remove {} skill", first.entity_name), agent);
        }
        SemanticChangeCode::SkillExecutableAppeared => {
            return scoped_title(&format!("update {} skill", first.entity_name), agent);
        }
        SemanticChangeCode::HookAdded
        | SemanticChangeCode::HookRemoved
        | SemanticChangeCode::HookChanged => "update hooks",
        SemanticChangeCode::InstructionChanged => "update project instructions",
        SemanticChangeCode::UnsupportedStateChanged => "update unsupported setup",
    };

    scoped_title(title, agent)
}

fn changed_surfaces_for_diff(diff: Option<&GraphDiff>) -> Vec<TimelineChangedSurface> {
    let Some(diff) = diff else {
        return Vec::new();
    };

    let mut surfaces = Vec::new();
    let mut semantic_source_paths = HashSet::new();

    for change in &diff.semantic_changes {
        let kind = timeline_surface_kind(change);
        let path = change
            .details
            .source_path
            .clone()
            .unwrap_or_else(|| "unknown".to_string());
        semantic_source_paths.insert(path.clone());
        let restorable = kind == "mcp_server" && path.ends_with(".mcp.json");
        surfaces.push(TimelineChangedSurface {
            kind,
            change_type: change.code.as_str().to_string(),
            path,
            entity_name: Some(change.entity_name.clone()),
            restorable,
            observe_only: !restorable,
            before: change.before.clone(),
            after: change.after.clone(),
        });
    }

    for change in &diff.raw_source_changes {
        if semantic_source_paths.contains(&change.source_path) {
            continue;
        }
        surfaces.push(TimelineChangedSurface {
            kind: "other".to_string(),
            change_type: format!("RAW_{}", change.status.to_uppercase()),
            path: change.source_path.clone(),
            entity_name: None,
            restorable: false,
            observe_only: true,
            before: None,
            after: None,
        });
    }

    surfaces
}

fn event_kind_for(
    previous: Option<&TimelineEntry>,
    diff: Option<&GraphDiff>,
) -> TimelineEntryEventKind {
    if previous.is_none() {
        return TimelineEntryEventKind::Baseline;
    }
    if let Some(diff) = diff {
        if diff.semantic_changes.is_empty() && diff.raw_source_changes.is_empty() {
            return TimelineEntryEventKind::Unchanged;
        }
    }
    TimelineEntryEventKind::SetupChanged
}

fn timeline_surface_kind(change: &SemanticChange) -> String {
    match change.entity_kind {
        EvidenceKind::McpServer => "mcp_server".to_string(),
        EvidenceKind::Skill => "skill".to_string(),
        EvidenceKind::Permission => "permission".to_string(),
        EvidenceKind::Hook => "hook".to_string(),
        EvidenceKind::EnvKey => "env_key".to_string(),
        EvidenceKind::Unsupported => "unsupported".to_string(),
        _ => "other".to_string(),
    }
}

fn restore_readiness_for(surfaces: &[TimelineChangedSurface]) -> TimelineRestoreReadiness {
    if surfaces.is_empty() {
        return TimelineRestoreReadiness::ObserveOnly;
    }
    let restorable = surfaces.iter().filter(|surface| surface.restorable).count();
    if restorable == surfaces.len() {
        TimelineRestoreReadiness::Full
    } else if restorable > 0 {
        TimelineRestoreReadiness::Partial
    } else {
        TimelineRestoreReadiness::ObserveOnly
    }
}

fn confidence_for(
    diff: Option<&GraphDiff>,
    surfaces: &[TimelineChangedSurface],
    diff_error: Option<&str>,
) -> TimelineConfidence {
    if diff_error.is_some() {
        return TimelineConfidence::Low;
    }
    if diff.is_none() {
        return TimelineConfidence::High;
    }
    if surfaces.iter().any(|surface| surface.path == "unknown") {
        return TimelineConfidence::Medium;
    }
    TimelineConfidence::High
}

fn confidence_reason_for(
    diff: Option<&GraphDiff>,
    surfaces: &[TimelineChangedSurface],
    diff_error: Option<&str>,
) -> String {
    if let Some(error) = diff_error {
        return format!("previous snapshot could not be diffed: {error}");
    }
    if diff.is_none() {
        return "first manual history baseline".to_string();
    }
    if surfaces.is_empty() {
        return "no semantic or raw source changes".to_string();
    }
    if surfaces.iter().any(|surface| surface.path == "unknown") {
        return "some changes lacked source path metadata".to_string();
    }
    "derived from snapshot graph diff".to_string()
}

fn priority_for_change(change: &SemanticChange) -> String {
    let code = change.code.as_str();
    if code.starts_with("MCP_") {
        return format!("0-{}", change.entity_name);
    }
    if code == "SKILL_EXECUTABLE_APPEARED" {
        return format!("1-{}", change.entity_name);
    }
    if code == "PERMISSION_WILDCARD_ADDED" {
        return format!("2-{}", change.entity_name);
    }
    if code.starts_with("ENV_KEY_") {
        return format!("4-{}", change.entity_name);
    }
    format!("5-{}", change.entity_name)
}

fn scoped_title(title: &str, agent: Option<AgentId>) -> String {
    match agent {
        Some(agent) => format!("{title} for {}", agent_label(agent)),
        None => title.to_string(),
    }
}

fn agent_label(agent: AgentId) -> &'static str {
    match agent {
        AgentId::ClaudeCode => "Claude Code",
        AgentId::Codex => "Codex",
        AgentId::Cursor => "Cursor",
        AgentId::Opencode => "OpenCode",
        AgentId::PiAgent => "Pi Agent",
        AgentId::Project => "Project",
        AgentId::Unknown => "Unknown",
    }
}

fn highlights_for_diff(diff: Option<&GraphDiff>) -> Vec<String> {
    let Some(diff) = diff else {
        return Vec::new();
    };

    let mut highlights: Vec<String> = diff
        .semantic_changes
        .iter()
        .take(5)
        .map(|change| format!("{}: {}", change.code.as_str(), change.entity_name))
        .collect();

    let remaining = 8usize.saturating_sub(highlights.len());
    highlights.extend(
        diff.raw_source_changes
            .iter()
            .take(remaining)
            .map(|change| format!("{}: {}", change.status, change.source_path)),
    );
    highlights
}